Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm

```diff
@@ -64,80 +64,77 @@
         return;
 
-    auto returnEmptyResult = ^() {
-        callback->call(@{ });
-    };
-
-    auto validateJSONAndReturnData = ^(id items, JSONOptionSet options, NSString **outExceptionString) {
-        if (!isValidJSONObject(items, options)) {
-            *outExceptionString = toErrorString(nil, @"items", @"it is not JSON-serializable");
-            return;
-        }
+    Vector<String> keysVector;
+    NSDictionary *keysWithDefaultValues = dynamic_objc_cast<NSDictionary>(items);
 
-        NSData *data = encodeJSONData(items, options);
-        if (!data) {
-            *outExceptionString = toErrorString(nil, @"items", @"it is not JSON-serializable");
-            return;
-        }
-
-        WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::StorageGet(page->webPageProxyIdentifier(), m_type, API::Data::createWithoutCopying(data)->dataReference()), [protectedThis = Ref { *this }, callback = WTFMove(callback)](std::optional<IPC::DataReference> dataReference, WebKit::WebExtensionContext::ErrorString error) {
-            if (error)
-                callback->reportError(error.value());
-            else
-                callback->call(parseJSON(API::Data::create(dataReference.value())));
-        }, extensionContext().identifier());
-    };
-
-    if ([items isKindOfClass:NSDictionary.class]) {
-        auto *keysWithDefaultValues = (NSDictionary *)items;
+    if (keysWithDefaultValues) {
         if (!keysWithDefaultValues.count) {
-            returnEmptyResult();
+            callback->call(@{ });
             return;
         }
 
-        validateJSONAndReturnData(keysWithDefaultValues, { }, outExceptionString);
-        return;
+        keysVector = makeVector<String>(keysWithDefaultValues.allKeys);
     }
 
-    if ([items isKindOfClass:NSArray.class]) {
-        auto *keys = (NSArray *)items;
+    if (NSArray *keys = dynamic_objc_cast<NSArray>(items)) {
         if (!keys.count) {
-            returnEmptyResult();
+            callback->call(@{ });
             return;
         }
 
-        validateJSONAndReturnData(keys, { JSONOptions::FragmentsAllowed }, outExceptionString);
-        return;
+        keysVector = makeVector<String>(keys);
     }
 
-    if ([items isKindOfClass:NSString.class]) {
-        auto *key = (NSString *)items;
+    if (NSString *key = dynamic_objc_cast<NSString>(items)) {
         if (!key.length) {
-            returnEmptyResult();
+            callback->call(@{ });
             return;
         }
 
-        validateJSONAndReturnData(key, { JSONOptions::FragmentsAllowed }, outExceptionString);
-        return;
+        keysVector = { key };
     }
+
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::StorageGet(page->webPageProxyIdentifier(), m_type, keysVector), [&, context, keysWithDefaultValues, protectedThis = Ref { *this }, callback = WTFMove(callback)](std::optional<String> dataJSON, WebKit::WebExtensionContext::ErrorString error) {
```

Comment:
`context` is not safe to capture. It is a local JS stack-frame essentially. We would need to use the global context and properly retain it with `JSRetainPtr`. Or get the global context from the `callback` object via `globalContext()`. Or use `parseJSON()` in CocoaHelpers (this is what I would do, since it avoids the indirection of `toNSObject()`.)